### ISSUE TITLE
Calling a function using JSX is producing different typed results vs calling it using raw JS

### ISSUE THREAD

**Issue by hadihallak** [microsoft/TypeScript#42836]
# Bug Report

Calling a function using JSX is producing different typed results vs calling it using raw JS
The event is typed in one but not the other.

I know that there are a lot of complications regarding making props polymorphic and typescript but I'm not sure why these are producing different results–shouldn't both events have the exact type?

### 🔎 Search Terms

polymorphism, react, as prop

### 🕗 Version & Regression Information
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about react and jsx

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.2.0-beta#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA3AB2MWUBeWcAwiZA1kwB44ABWJhUAPjgBvGgEgBAQXRYAHswYATdNjwwAdADFOEgBRgAXDJSoryuAF84AMkxUD3cBD5NREcUqSLiJiqACUVrr4+lEwAKIANlgg-DA0DnS43qjwACLAAG6R7kacAgBSAMoAGvoAkkxQjKjAuInJqagA2mSahWQAulIAvHCmYXDDUrJylDAArlAMcAzzCQnpmdnwisV6pRU19Y3Nre0pTN1kyIMjYxNTMvJzi8ur65s0apCwcFkMOTgACF5jAYN5JmNTONJtMHBM0CgGABPBHoTy8VICaRwArIJrIJgAfisFCwmjIcAAPuQAEYJeZYSkOCR0GgAenZcAA6gALZFwAAGWAKqUFcGA6BgyLA5LgvJYWBo-0BcVFTDqqAAKjK5aMQWDvKYcWg7AAaODeTgJVoAaysphFqQecMcYTonLgtNBKwg8AVlGV2zgatSADk-TrZZpIQIDeDlmhhtJFE4rTbcLbk471TAXTIHOyJEA)

### 💻 Code

```ts
import * as React from 'react'

interface Component< Props> {
	<As extends React.FC>(p: { as: As } & React.ComponentProps<As> & Props): React.ReactElement
}

const Div: React.FC<JSX.IntrinsicElements['div']> = () => {
	return null
}

const A: React.FC<JSX.IntrinsicElements['a']> = () => {
	return null
}

export const Button = ((() => {}) as any) as Component<{ variant?: 'red' | 'blue' }>



// Why `event` is typed here
const EventIsTyped = Button({ as: A, onClick: (event) => {} })

// but not here
const EventNotTyped = <Button as={A} onClick={(event) => {}/>
```

### 🙁 Actual behavior

The JSX call is producing an untyped event vs to calling the function using normal js
- in `EventIsTyped` the event is typed as `React.MouseEvent<HTMLAnchorElement, MouseEvent>`
- in `EventNotTyped` the event is typed as implicit any even though the onClick attr has the correct type

### 🙂 Expected behavior

if the js function call is producing a typed event, I would expect that the event would be typed in the jsx call and vice versa 
so in short i would expect that `Button({ as: A, onClick: (event) => {} })` === `<Button as={A} onClick={(event) => {}/>` when it comes to how they're typed

**Comment by RyanCavanaugh** [comment]
See #14729; JSX is not desugared into a function call for typechecking purposes for a variety of reasons.
